src/gap.c

```diff
@@ -1374,75 +1341,91 @@ Obj FuncSetUserHasQuit( Obj Self, Obj value)
+	 " There is already a timeout running", 0, 0); */
+      SyStopAlarm((UInt *)&curr_seconds, (UInt *)&curr_nanoseconds);
+      curr_microseconds = curr_nanoseconds/1000;
+      if (iseconds > curr_seconds || (iseconds == curr_seconds && imicroseconds > curr_microseconds)) {
+	  /* The existing timeout will go off before we would so don't bother to set a timeout
+	   just reset the existing one and call the function */
+	  iseconds = curr_seconds;
+	  imicroseconds = curr_microseconds;
+	  newJumpBuf = 0;
+	} else {
+	/* We would time out before the other function, so we need to
+	   set our timeout, but remember to reset things later */
+	mayNeedToRestore = 1;
+	newJumpBuf = 1;
       }
-    else 
-      plain_args = args;
-    if (SyAlarmRunning)
-      ErrorMayQuit("CALL_WITH_TIMEOUT cannot currently be nested except via break loops."
-		   " There is already a timeout running", 0, 0);
-    if (NumAlarmJumpBuffers >= MAX_TIMEOUT_NESTING_DEPTH-1)
-      ErrorMayQuit("Nesting depth of timeouts via break loops limited to %i", MAX_TIMEOUT_NESTING_DEPTH, 0L);
-    currLVars = TLS(CurrLVars);
-    currStat = TLS(CurrStat);
-    recursionDepth = TLS(RecursionDepth);
-    if (sySetjmp(AlarmJumpBuffers[NumAlarmJumpBuffers++])) {
-      /* Timeout happened */
-      TLS(CurrLVars) = currLVars;
-      TLS(PtrLVars) = PTR_BAG(TLS(CurrLVars));
-      TLS(PtrBody) = (Stat*)PTR_BAG(BODY_FUNC(CURR_FUNC));
-      TLS(CurrStat) = currStat;
-      TLS(RecursionDepth) = recursionDepth;
-      res = Fail;
-    } else {
-      SyInstallAlarm( INT_INTOBJ(seconds), 1000*INT_INTOBJ(microseconds));
-      switch (LEN_PLIST(plain_args)) {
-      case 0: result = CALL_0ARGS(func);
-        break;
-      case 1: result = CALL_1ARGS(func, ELM_PLIST(plain_args,1));
-        break;
-      case 2: result = CALL_2ARGS(func, ELM_PLIST(plain_args,1),
-                                  ELM_PLIST(plain_args,2));
-        break;
-      case 3: result = CALL_3ARGS(func, ELM_PLIST(plain_args,1),
-                                  ELM_PLIST(plain_args,2), ELM_PLIST(plain_args,3));
-        break;
-      case 4: result = CALL_4ARGS(func, ELM_PLIST(plain_args,1),
-                                  ELM_PLIST(plain_args,2), ELM_PLIST(plain_args,3),
-                                  ELM_PLIST(plain_args,4));
-        break;
-      case 5: result = CALL_5ARGS(func, ELM_PLIST(plain_args,1),
-                                  ELM_PLIST(plain_args,2), ELM_PLIST(plain_args,3),
-                                  ELM_PLIST(plain_args,4), ELM_PLIST(plain_args,5));
-        break;
-      case 6: result = CALL_6ARGS(func, ELM_PLIST(plain_args,1),
-                                  ELM_PLIST(plain_args,2), ELM_PLIST(plain_args,3),
-                                  ELM_PLIST(plain_args,4), ELM_PLIST(plain_args,5),
-                                  ELM_PLIST(plain_args,6));
-        break;
-      default: result = CALL_XARGS(func, plain_args);
+    }
```

Comment:
Hrm, it's really a bit difficult to see what changed regarding the timer code, with the `CallFuncList` change mixed in. So, yeah, having `CallFuncList` added in a separate prior commit would be really helpful for reviewers.